sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala

```diff
@@ -299,11 +292,26 @@ private[orc] class OrcOutputWriter(
   }
 
   override def close(): Unit = {
-    if (recordWriterInstantiated) {
+    try {
+      OrcUtils.addSparkVersionMetadata(getOrCreateInternalWriter())
+    } catch {
+      case NonFatal(e) => log.warn(e.toString, e)
+    }
+    recordWriter.close(Reporter.NULL)
+  }
+
+  private def getOrCreateInternalWriter(): Writer = {
+    val writerField = recordWriter.getClass.getDeclaredField("writer")
+    writerField.setAccessible(true)
+    var writer = writerField.get(recordWriter).asInstanceOf[Writer]
+    if (writer == null) {
```

Comment:
do we need to consider multi-thread here?

Reply:
org.apache.hadoop.hive.ql.io.orc.OrcOutputFormat.OrcRecordWriter init `writer` while writing the first element without synchronizing and we call `write` and `close` serially. I guess it's safe here